File: services/agent-api/app/api/routes/reliability.py
"""Reliability diagnostics and user-controlled data-rights operations."""

import uuid

from fastapi import APIRouter

from app.schemas.common import err, ok
from app.schemas.reliability import (
    CompanionDeletionCreateRequest,
    CompanionDeletionExecuteRequest,
    DataRightsDryRunRequest,
)
from app.services import (
    data_rights_deletion_service,
    data_rights_export_service,
    reliability_diagnostics_service,
)


router = APIRouter(tags=["Reliability"])


@router.get("/companions/{companion_id}/reliability-diagnostics")
def get_reliability_diagnostics(companion_id: str):
    try:
        return ok(reliability_diagnostics_service.get_reliability_diagnostics(uuid.UUID(companion_id)))
    except (ValueError, TypeError) as exc:
        return err("RELIABILITY_SCOPE_INVALID", str(exc))


@router.post("/companions/{companion_id}/data-rights/dry-run")
def dry_run_data_rights(companion_id: str, body: DataRightsDryRunRequest):
    try:
        return ok(reliability_diagnostics_service.get_data_rights_dry_run(uuid.UUID(companion_id), body.operation, body.target_id))
    except (ValueError, TypeError) as exc:
        return err("DATA_RIGHTS_DRY_RUN_REJECTED", str(exc))


@router.post("/companions/{companion_id}/data-rights/export")
def export_companion_data(companion_id: str):
    try:
        return ok(
            data_rights_export_service.export_companion_data(
                uuid.UUID(companion_id)
            )
        )
    except (ValueError, TypeError) as exc:
        return err("COMPANION_DATA_EXPORT_REJECTED", str(exc))


@router.get("/companions/{companion_id}/data-rights/deletion-request")
def get_companion_deletion_request(companion_id: str):
    try:
        return ok(
            data_rights_deletion_service.get_companion_deletion_request(
                uuid.UUID(companion_id)
            )
        )
    except (ValueError, TypeError) as exc:
        return err("COMPANION_DELETION_STATUS_REJECTED", str(exc))


@router.get("/data-rights/deletion-requests/{request_id}")
def get_deletion_request(request_id: str):
    try:
        return ok(
            data_rights_deletion_service.get_deletion_request(
                uuid.UUID(request_id)
            )
        )
    except (ValueError, TypeError) as exc:
        return err("COMPANION_DELETION_STATUS_REJECTED", str(exc))


@router.post("/companions/{companion_id}/data-rights/deletion-requests")
def create_companion_deletion_request(
    companion_id: str,
    body: CompanionDeletionCreateRequest,
):
    try:
        return ok(
            data_rights_deletion_service.create_deletion_request(
                uuid.UUID(companion_id),
                confirmation_name=body.confirmation_name,
                skip_recovery_window=body.skip_recovery_window,
                export_choice=body.export_choice,
                idempotency_key=body.idempotency_key,
            )
        )
    except (ValueError, TypeError) as exc:
        return err("COMPANION_DELETION_REJECTED", str(exc))


@router.post("/data-rights/deletion-requests/{request_id}/restore")
def restore_companion_deletion_request(request_id: str):
    try:
        return ok(
            data_rights_deletion_service.restore_deletion_request(
                uuid.UUID(request_id)
            )
        )
    except (ValueError, TypeError) as exc:
        return err("COMPANION_DELETION_RESTORE_REJECTED", str(exc))


@router.post("/data-rights/deletion-requests/{request_id}/execute")
def execute_companion_deletion_request(
    request_id: str,
    body: CompanionDeletionExecuteRequest,
):
    try:
        return ok(
            data_rights_deletion_service.execute_deletion_request(
                uuid.UUID(request_id),
                allow_before_due=body.allow_before_due,
            )
        )
    except (ValueError, TypeError) as exc:
        return err("COMPANION_DELETION_EXECUTION_REJECTED", str(exc))

File: apps/web/lib/api/reliability.ts
import { api } from "./client";

export type ReliabilityDomain = {
  key: string;
  label: string;
  status: "healthy" | "attention" | "blocked" | "available" | "unavailable";
  total: number;
  active?: number;
  terminal?: number;
  failed?: number;
  stuck?: number;
  status_counts: Record<string, number>;
  safe_reason?: string;
};

export type ReliabilityDiagnostics = {
  contract_version: string;
  companion_id: string;
  companion_status: string;
  generated_at: string;
  overall_status: "healthy" | "attention" | "blocked";
  runtime_domains: ReliabilityDomain[];
  state_domains: ReliabilityDomain[];
  validation_matrix: Array<{
    key: string;
    label: string;
    scope: "companion";
    all_time_evidence_count: number;
    last_30d_evidence_count: number;
    evidence_sources: string[];
    coverage_status: "evidence_available" | "no_evidence";
    quality_conclusion: "snapshot_only_no_long_term_claim";
  }>;
  quality: { open_bad_cases: number; failed_regression_results: number };
  safety: {
    active_hard_stops: number;
    revoked_channel_bindings: number;
    pending_shared_reviews: number;
    memory_reranker_policy_mode: string;
    presence_bandit_policy_mode: string;
    observer_auto_speaker: boolean;
  };
  capabilities: Record<string, boolean | string>;
  retention_boundaries: Record<string, string>;
  content_disclosure: string;
};

export type DataRightsOperation = "export" | "forget_memory" | "archive_companion" | "disconnect_channels" | "revoke_channels" | "permanent_delete";

export type DataRightsDryRun = {
  contract_version: string;
  companion_id: string;
  operation: DataRightsOperation;
  dry_run: true;
  executed: false;
  supported_write_path: boolean;
  irreversible: boolean;
  separate_authorization_required: boolean;
  ready_for_explicit_execution: boolean;
  blockers: string[];
  effect_summary: string;
  affected_counts: Record<string, number>;
  retained_evidence: string[];
  review_gates: string[];
  content_disclosure: "counts_only";
};

export type CompanionDeletionRequest = {
  contract_version: string;
  id: string;
  companion_id: string | null;
  companion_display_name: string | null;
  status: "trash" | "purging" | "completed" | "restored" | "failed";
  deletion_mode: "recovery_window" | "immediate";
  current_stage: string;
  requested_at: string;
  purge_after: string;
  completed_at: string | null;
  restored_at: string | null;
  backup_delete_due_at: string | null;
  affected_counts: Record<string, number>;
  deleted_counts: Record<string, number>;
  failure_code: string | null;
  failure_stage: string | null;
  can_restore: boolean;
  can_retry: boolean;
  content_disclosure: "counts_and_safe_status_only";
};

export type CompanionDataExport = {
  contract_version: "companion-data-export.v1";
  exported_at: string;
  owner_id: string;
  companion: {
    id: string;
    name: string;
    [key: string]: unknown;
  };
  sections: Record<string, Array<Record<string, unknown>>>;
  manifest: {
    format: "json";
    section_counts: Record<string, number>;
    excluded: string[];
    sha256: string;
  };
};

export const getReliabilityDiagnostics = (companionId: string) =>
  api.get<ReliabilityDiagnostics>(`/companions/${companionId}/reliability-diagnostics`);

export const dryRunDataRights = (companionId: string, operation: DataRightsOperation, targetId?: string) =>
  api.post<DataRightsDryRun>(`/companions/${companionId}/data-rights/dry-run`, { operation, target_id: targetId || undefined });

export const exportCompanionData = (companionId: string) =>
  api.post<CompanionDataExport>(
    `/companions/${companionId}/data-rights/export`,
  );

export const getCompanionDeletionRequest = (companionId: string) =>
  api.get<CompanionDeletionRequest | null>(`/companions/${companionId}/data-rights/deletion-request`);

export const getDeletionRequest = (requestId: string) =>
  api.get<CompanionDeletionRequest>(`/data-rights/deletion-requests/${requestId}`);

export const createCompanionDeletionRequest = (
  companionId: string,
  input: {
    confirmation_name: string;
    skip_recovery_window: boolean;
    export_choice: "skip" | "completed";
    idempotency_key: string;
  },
) =>
  api.post<CompanionDeletionRequest>(
    `/companions/${companionId}/data-rights/deletion-requests`,
    input,
  );

export const restoreCompanionDeletionRequest = (requestId: string) =>
  api.post<CompanionDeletionRequest>(
    `/data-rights/deletion-requests/${requestId}/restore`,
  );

export const executeCompanionDeletionRequest = (requestId: string) =>
  api.post<CompanionDeletionRequest>(
    `/data-rights/deletion-requests/${requestId}/execute`,
    { allow_before_due: true },
  );
